### PR TITLE
Support offline private tally in ledger

### PR DESCRIPTION
**Pull request by eugene-babichenko**

### CODE REVIEW

File: chain-impl-mockchain/src/certificate/vote_tally.rs [hunk 1]
```diff
@@ -40,8 +40,14 @@ pub enum TallyProof {
...
-pub struct TallyDecryptShares {
-    inner: Box<[Box<[TallyDecryptShare]>]>,
+pub struct PrivateTallyDecrypted {
+    inner: Box<[PrivateTallyDecryptedProposal]>,
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct PrivateTallyDecryptedProposal {
+    pub decrypt_shares: Box<[TallyDecryptShare]>,
+    pub tally_result: Box<[u64]>,
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/certificate/vote_tally.rs:50]: Shouldn't the result be included just once? It does not need to be duplicated per each share.

**Review by mzabaluev — CHANGES_REQUESTED** [review]

File: chain-impl-mockchain/src/certificate/vote_tally.rs (hunk 1 above)
Comment (mzabaluev) [reply chain-impl-mockchain/src/certificate/vote_tally.rs:50]: Sorry, confused about the array's dimensions.

**Review by mzabaluev — APPROVED** [review]

File: chain-impl-mockchain/src/certificate/vote_tally.rs [hunk 2]
```diff
@@ -90,20 +102,25 @@ impl VoteTally {
...
+                    |bb, proposal| {
                         // Shares per proposal, n_members x n_options
-                        let n_members = s.len().try_into().unwrap();
+                        let n_members = proposal.decrypt_shares.len().try_into().unwrap();
                         if n_members == 0 {
                             bb.u8(0).u8(0)
                         } else {
-                            let n_options = s[0].options().try_into().unwrap();
+                            let n_options = proposal.tally_result.len().try_into().unwrap();
                             bb.u8(n_members)
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply chain-impl-mockchain/src/certificate/vote_tally.rs:115]: Strictly speaking, it doesn't contradict the tallying algorithm. But overall this is an interesting idea.

[the remark this answers is not in the document]
Comment (mzabaluev) [reply chain-impl-mockchain/src/certificate/vote_tally.rs:115]: The treasury voting protocol paper stipulates that decryption shares from every committee member are either provided or recovered by posting shares of their communication keys. Not sure if the recovered shares need to be separately posted on the blockchain, though; so let it stay this way.

**Review by mzabaluev — APPROVED** [review]

**Pull request merged**